xla/stream_executor/cuda/cuda_dnn.cc

```diff
@@ -8368,10 +7996,105 @@ class CudnnExecutionPlanRunner<void(Args...)>
   mutable int64_t initial_offset_ = 0;
   int64_t offset_increment_ = 0;
   int64_t rng_seed_;
-  bool is_flash_attention_;
 };
 #endif  // CUDNN_VERSION >= 8100
 
+template <typename Sig>
+class CudnnGraphRunner;
+// An OpRunner implemented by a cuDNN frontend graph.
+//
+// This is the class holding the implementation of ToString, GetWorkspaceSize,
+// and operator() for use by the cudnn frontend op runners.
+template <typename... Args>
+class CudnnGraphRunner<void(Args...)> : public dnn::OpRunner<void(Args...)> {
+ private:
+  using Graph = cudnn_frontend::graph::Graph;
+  using Tensor_attributes = cudnn_frontend::graph::Tensor_attributes;
+
+ public:
+  std::string ToString() const override { return graph_.Graph().print(); }
+
+  size_t GetWorkspaceSize() const override {
+    return graph_.Graph().get_workspace_size();
+  }
+
+  absl::StatusOr<dnn::AlgorithmDesc> ToAlgorithmDesc() const override {
+    return absl::InternalError(
+        "Unexpected call to CudnnGraphRunner::ToAlgorithmDesc");
+  }
+
+  absl::Status operator()(Stream* stream, dnn::ProfileResult* profile_result,
+                          DeviceMemoryBase scratch_memory,
+                          Args... inputs) const override {
+    if (static_cast<internal::StreamExecutorInterface*>(parent_) !=
+        stream->parent()->implementation()) {
+      return tsl::errors::Internal(
+          "CudnnExecutionPlanRunner cached across multiple StreamExecutors.");
+    }
+    CudnnHandle handle = cudnn_->GetHandle(parent_, stream);
+    std::unordered_map<int64_t, void*> variant_pack;
+    std::vector<void*> vec = {inputs.opaque()...};
+    for (int i = 0; i < uids_.size(); ++i) {
+      if (uids_[i].has_value()) {
+        variant_pack[*uids_[i]] = vec[i];
+      }
+    }
+
+    for (auto& [uid, value] : scalars_) {
+      variant_pack[uid] = const_cast<void*>(value.ToVoidPointer());
+    }
+
+    if (dropout_rng_offset_increment_ > 0) {
+#if CUDNN_VERSION >= 8800
+      variant_pack[D_SEED_ID] = (void*)&dropout_rng_seed_;
+      current_dropout_rng_offset_ += dropout_rng_offset_increment_;
+      variant_pack[D_OFFSET_ID] = (void*)&current_dropout_rng_offset_;
+#else
+      return absl::UnimplementedError(
+          "Cudnn dropout offset and seed are only supported with Cudnn >= "
+          "8.8.0");
+#endif  // CUDNN_VERSION >= 8800
+    }
+
+    RETURN_IF_CUDNN_FRONTEND_ERROR(graph_.Graph().execute(
+        handle.handle(), variant_pack, scratch_memory.opaque()));
+
+    return tsl::OkStatus();
+  }
+
+  static absl::StatusOr<CudnnGraphRunner> Create(
+      GpuExecutor* parent, CudnnAccess* cudnn, CudnnGraph graph,
+      int64_t dropout_rng_seed, int64_t dropout_rng_offset,
+      std::unordered_map<int64_t, ScalingParam> scalars,
+      std::vector<std::optional<int64_t>> uids) {
+    return {{parent, cudnn, graph, dropout_rng_seed, dropout_rng_offset,
```

Comment:
Would `std::move(graph)` be better here?

Reply:
This is now 

```
    return std::move(CudnnGraphRunner(parent, cudnn, graph, dropout_rng_seed,
                                      dropout_rng_offset, uids));
```

which is a performance pessimization because it prevents return value optimization. Please move the `graph`, not the `CudnnGraphRunner`.

Reply:
Fixed it. Sorry for the confusion.